native/data_dypes.md

```diff
@@ -0,0 +1,102 @@
+
+ - Persistent / Rooted Variable
+ - Scoped / Temp Variable
+
+`Node` is not expected to manage the memory for JS Engine variable for any of the given options above. One of the [verified](https://github.com/jxcore/jxcore/blob/master/doc/native/Embedding_Basics.md) assumptions show that; a void ptr can point to a native JS Engine variable without dealing with its memory management.
+
+`void *data_;` -> JS Engine Variable
+
+Similar to local C,C++ variables, if `data_` holds a scope variable, it's value is not expected to be available after function's scope. Persistent or rooted variables are not restricted with function scope. However, they need to be freed manually.
+
+V8 API mentions a `weak-able` variable. This type of variable is not expected to be GC'ed by `JS engine` until there is no Javascript side variable referencing to it. 
+
+`JS engines` by default do not GC any active variable. As long as there is one another (non GC-able) variable holding a reference to the subject variable, it should be reachable. However, V8 also doesn't guarantee that the `native representation` of non-persistent but weakened variable will persist. As a result, from the native side scope/data point of view, `weak-able` variables do not bring a new condition or challenge. All the mature `JS Engines` provide somehow a similar functionality.
+
+V8 engine (similar to other mature `JS engines`) has a GC event mechanism on C,C++ side to tell a weak variable has finally / or about to be garbage collected. Although the timing of the GC needs a separate discussion, it has not much to do with data type abstraction subject. 
+
+`JS Engine` uses a thread, context and or scope variable for each interaction with its API. `Node` also stores an environment / common object per application or thread instance. `Node` and / or `JS Engine` already manages the memory of this global variable. Based on similarly verified previous assumption, void ptr pointing the underlying global variable may serve the same purpose.
+
+`void *env_;` -> Env, commons, Isolate, JSContext ...
+
+Until this point, three variables of the expected interface type is discussed.
+
+```
+  void *env_;
+  void *data_;
+  JSTypes type_;
+```
+
+Before closing up, some other pieces may require an additional attention.
+
+ - Underlying implementation may require wrapping either `env_` or `data_` prior to delivering it to addon land. ie. engine specific scenarios for specialized type implementations.
+
+ - It may not be clear on the engine side whether a soft wrapped (void ptr) data type is rooted/persistent or not.
+
+For the last two items and more, `uint_8 flags_` provides an implementation specific information for the interface itself.
+
+Initially suggested interface is given below
+
+```
+struct JSValue {
+  void *env_;
+  void *data_;
+  uint_8 flags_; // implementation specific
+  JSTypes type_;
```

Comment:
You might as well make `flags_` a uintptr_t and fold `type_` into it.  ABI requirements means it's going to get padded up to word size anyway.